synapse/storage/util/id_generators.py

```diff
@@ -722,6 +738,9 @@ def get_current_token_for_writer(self, instance_name: str) -> int:
         # persisted up to position. This stops Synapse from doing a full table
         # scan when a new writer announces itself over replication.
         with self._lock:
+            if self._instance_name == instance_name:
+                return self._return_factor * self._max_position_of_local_instance
```

Comment:
Do we still need the lock here? I guess the lock guards `_max_position_of_local_instance`... but I'm not sure what that buys us?

Reply:
We lock on all functions, as they can be called from DB threads. It's *probably* safe to move this particular thing out of the lock, but that's only true because it's a trivial expression (only `_max_position_of_local_instance` can change). With locks I'd kinda prefer to keep them consistent and just have everything in them